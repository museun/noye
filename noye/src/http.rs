#![cfg_attr(debug_assertions, allow(dead_code,))]
use std::collections::HashMap;

/// so no query-strings can be used with get_json
#[derive(serde::Serialize)]
pub struct NoQuery;

pub async fn get_body<C>(
    client: &surf::Client<C>,
    url: &str,
    headers: &[(&'static str, &str)],
) -> anyhow::Result<Vec<u8>>
where
    C: surf::middleware::HttpClient,
{
    let mut req = client.get(url);
    for (k, v) in headers {
        req = req.set_header(k, v);
    }
    req.recv_bytes()
        .await
        .map_err(|err| anyhow::anyhow!("cannot get url '{}': {}", url, err))
}

pub async fn get_json<'a, T, C, Q>(
    client: &surf::Client<C>,
    url: &'a str,
    query: &'a Q,
    headers: &[(&'static str, &str)],
) -> anyhow::Result<T>
where
    for<'de> T: serde::Deserialize<'de>,
    C: surf::middleware::HttpClient,
    Q: serde::Serialize + 'a,
{
    let mut req = client.get(url);
    if let Some(query) = query.into() {
        req = req.set_query(query).unwrap();
    }
    for (k, v) in headers {
        req = req.set_header(k, v);
    }
    req.recv_json()
        .await
        .map_err(|err| anyhow::anyhow!("cannot get url '{}': {}", url, err))
}

pub async fn head<C>(client: &surf::Client<C>, url: &str) -> anyhow::Result<surf::Response>
where
    C: surf::middleware::HttpClient,
{
    client
        .head(url)
        .await
        .map_err(|err| anyhow::anyhow!("failed to do HEAD on {}: {}", url, err))
}

pub struct Url {
    domain: String,
    path: String,
    query: HashMap<String, String>,
}

impl std::convert::TryFrom<&str> for Url {
    type Error = ();
    fn try_from(data: &str) -> std::result::Result<Self, Self::Error> {
        let url = data.parse::<http::Uri>().map_err(|_| ())?;
        let query_map = url
            .query()
            .into_iter()
            .flat_map(|s| s.split('&'))
            .filter_map(|s| {
                let mut s = s.split('=').map(ToString::to_string);
                let (k, v) = (s.next()?, s.next()?);
                (k, v).into()
            });

        Ok(Self {
            domain: url.authority().ok_or_else(|| ())?.host().to_string(),
            path: url.path().to_string(),
            query: query_map.collect(),
        })
    }
}

impl Url {
    pub fn domain(&self) -> &str {
        &self.domain
    }

    pub fn path(&self) -> &str {
        &self.path
    }

    pub fn query<Q: ?Sized>(&self, key: &Q) -> Option<&str>
    where
        Q: std::hash::Hash + Eq,
        String: std::borrow::Borrow<Q>,
    {
        self.query.get(key).map(|s| s.as_str())
    }
}
