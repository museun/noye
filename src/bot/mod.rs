pub use crate::irc::Message;

mod context;
pub use context::Context;

mod resolver;
pub use resolver::Resolver;

mod responder;
pub use responder::Responder;

mod state;
pub use state::State;

mod writer;
pub use writer::Writer;

mod handler;
pub use handler::{AnyhowFut, CommandsMap, Handler, PassivesList};

mod runner;
pub use runner::Runner;
