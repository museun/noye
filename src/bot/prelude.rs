#[doc(inline)]
pub use crate::command::{Context, Dispatcher, Event, Noye};

#[doc(inline)]
pub use template::*;

#[doc(inline)]
pub use crate::config::{self, ApiKey, Config};

#[doc(inline)]
pub use crate::irc::{Command, Target};

#[doc(inline)]
pub use crate::matches::Matches;

#[doc(inline)]
pub use crate::format::{CommaSeparated, FileSize, Iso8601, Timestamp};

use futures::prelude::*;
pub async fn concurrent_map<Iter, Item, Func, Fut, Ok, Err, Hint>(
    namespace: &str,
    hint: Hint,
    input: Iter,
    func: Func,
) -> impl Stream<Item = Ok>
where
    Iter: IntoIterator<Item = Item>,
    Func: Copy + FnOnce(Item) -> Fut,
    Fut: TryFuture<Ok = Ok, Error = Err>,
    Err: std::fmt::Display,
    Hint: Into<Option<usize>>,
{
    let (buf_tx, buf_rx) = tokio::sync::mpsc::unbounded_channel();
    stream::iter(input)
        .for_each_concurrent(hint, |item| {
            let buf_tx = buf_tx.clone();
            async move {
                if let Ok(ok) = func(item)
                    .inspect_err(|err| log::warn!("{} produced an error: {}", namespace, err))
                    .await
                {
                    let _ = buf_tx.send(ok);
                }
            }
        })
        .await;
    drop(buf_tx); // drop so the stream will end
    buf_rx
}
