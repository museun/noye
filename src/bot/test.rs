pub use crate::bot::prelude::*;
pub use crate::irc::*;

pub use futures::prelude::*;
pub use tokio::sync::mpsc;

pub fn say_template<T: Template>(context: &Context, template: T) -> String {
    let data = crate::command::resolve_template(template).unwrap();
    match context.target().unwrap() {
        Target::Channel(target) | Target::Private(target) => {
            format!("PRIVMSG {} :{}", target, data)
        }
    }
}

pub fn reply_template<T: Template>(context: &Context, template: T) -> String {
    let data = crate::command::resolve_template(template).unwrap();
    match context.target().unwrap() {
        Target::Channel(target) => format!(
            "PRIVMSG {} :{}: {}",
            target,
            context.nick().expect("nick must be attached to message"),
            data
        ),
        Target::Private(target) => format!("PRIVMSG {} :{}", target, data),
    }
}

// TODO allow for asserting on errors
pub fn check_error<F, Fut>(func: F, input: Context)
where
    F: Copy + FnOnce(Context, Noye) -> Fut,
    Fut: Future<Output = anyhow::Result<()>>,
{
    let (tx, mut rx) = mpsc::channel(32);
    let noye = Noye::new(tx);
    tokio_test::block_on(async move {
        func(input, noye).await.unwrap_err();
        assert!(rx.next().await.is_none())
    });
}

pub fn check<F, Fut>(func: F, input: Context, mut output: Vec<&str>)
where
    F: Copy + FnOnce(Context, Noye) -> Fut,
    Fut: Future<Output = anyhow::Result<()>>,
{
    let (tx, mut rx) = mpsc::channel(32);
    let noye = Noye::new(tx);

    tokio_test::block_on(async move {
        if let Err(err) = func(input.clone(), noye).await {
            panic!("failed to run '{}' on '{:#?}'", err, input.message());
        }
        let mut index = 0_usize;
        while let Some(msg) = rx.next().await {
            if output.is_empty() {
                panic!("got input: ({}) '{}' but output was empty", index, msg);
            }
            let left = output.remove(0);
            assert_eq!(
                left,
                msg,
                "expected at output pos: {}. '{}' != '{}'",
                index,
                left.escape_debug(),
                msg.escape_debug()
            );
            index += 1;
        }
    });
}
