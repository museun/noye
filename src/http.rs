use anyhow::Context as _;

pub fn new_client() -> reqwest::Client {
    reqwest::Client::new()
}

#[derive(serde::Serialize)]
pub struct NoQuery;

#[allow(dead_code)]
pub async fn get_body(
    client: reqwest::Client,
    url: &str,
    headers: &[(&'static str, &str)],
) -> anyhow::Result<String> {
    let mut req = client.get(url);
    for &(k, v) in headers {
        req = req.header(k, v);
    }

    client
        .execute(req.build()?)
        .await
        .with_context(|| format!("cannot get url '{}'", url))?
        .error_for_status()
        .with_context(|| format!("cannot get url '{}'", url))?
        .text()
        .await
        .with_context(|| format!("cannot get body for '{}'", url))
}

pub async fn get_json<'a, T, Q>(
    client: reqwest::Client,
    url: &'a str,
    query: &'a Q,
    headers: &[(&'static str, &str)],
) -> anyhow::Result<T>
where
    for<'de> T: serde::Deserialize<'de>,
    Q: serde::Serialize + 'a,
{
    let mut req = client.get(url);
    if let Some(query) = query.into() {
        req = req.query(query);
    }

    for &(k, v) in headers {
        req = req.header(k, v);
    }

    client
        .execute(req.build()?)
        .await
        .with_context(|| format!("cannot get url '{}'", url))?
        .error_for_status()
        .with_context(|| format!("cannot get url '{}'", url))?
        .json()
        // .text()
        .await
        // .map(|s| {
        //     eprintln!("got json from: {}", url);
        //     let v: serde_json::Value = serde_json::from_str(&s).unwrap();
        //     std::fs::write(
        //         format!(
        //             "{}.json",
        //             url.replace("http://", "")
        //                 .replace("https://", "")
        //                 .replace('/', "_")
        //                 .replace(".", "_")
        //         ),
        //         serde_json::to_string_pretty(&v).unwrap(),
        //     )
        //     .unwrap();
        //     serde_json::from_value(v).unwrap()
        // })
        .with_context(|| format!("cannot get json for '{}'", url))
}

pub async fn head(client: reqwest::Client, url: &str) -> anyhow::Result<reqwest::Response> {
    let req = client.head(url).build()?;
    client
        .execute(req)
        .await
        .with_context(|| format!("cannot get HEAD for '{}'", url))?
        .error_for_status()
        .with_context(|| format!("cannot get url '{}'", url))
}
