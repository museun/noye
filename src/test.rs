#![allow(dead_code)]
pub use crate::{responses::*, *};
pub use futures::prelude::*;

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Default, Clone)]
pub struct YamlResponder {
    pub responses: Arc<std::sync::Mutex<Vec<Response>>>,
}

impl YamlResponder {
    pub fn len(&self) -> usize {
        self.responses.lock().unwrap().len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn get_say<T>(&self) -> T
    where
        for<'de> T: Deserialize<'de>,
    {
        match self.remove() {
            Response::Say(say) => serde_yaml::from_str(&say).unwrap(),
            d => panic!("expected 'say' got {}", resp_kind(d)),
        }
    }

    pub fn get_reply<T>(&self) -> T
    where
        for<'de> T: Deserialize<'de>,
    {
        match self.remove() {
            Response::Reply(reply) => serde_yaml::from_str(&reply).unwrap(),
            d => panic!("expected 'reply' got {}", resp_kind(d)),
        }
    }

    pub fn get_raw(&self) -> String {
        match self.remove() {
            Response::Raw(raw) => raw,
            d => panic!("expected 'raw' got {}", resp_kind(d)),
        }
    }

    pub fn expect_empty(&self) {
        assert!(
            self.is_empty(),
            "expected no responses, got: {}",
            self.len()
        )
    }

    fn remove(&self) -> Response {
        let responses = &mut *self.responses.lock().unwrap();
        assert!(!responses.is_empty(), "expected a response");
        responses.remove(0)
    }
}

impl Responder for YamlResponder {
    fn say<T>(&mut self, _: Context, template: T) -> AnyhowFut<'_>
    where
        T: template::Template + Send + Sync + 'static,
        T: Serialize,
    {
        Box::pin(async move {
            self.responses
                .lock()
                .unwrap()
                .push(Response::Say(serde_yaml::to_string(&template).unwrap()));
            Ok(())
        })
    }

    fn reply<T>(&mut self, _: Context, template: T) -> AnyhowFut<'_>
    where
        T: template::Template + Send + Sync + 'static,
        T: Serialize,
    {
        Box::pin(async move {
            self.responses
                .lock()
                .unwrap()
                .push(Response::Reply(serde_yaml::to_string(&template).unwrap()));
            Ok(())
        })
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Response {
    Say(String),
    Reply(String),
    Raw(String),
}

fn resp_kind(response: Response) -> &'static str {
    match response {
        Response::Reply(..) => "reply",
        Response::Raw(..) => "raw",
        Response::Say(..) => "say",
    }
}

pub fn set_snapshot_path() {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path(get_snapshot_path());
    settings.bind_to_thread();
}

pub fn get_snapshot_path() -> std::path::PathBuf {
    std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("snapshots")
}

pub struct TestEnv {
    responder: YamlResponder,
    data: String,
    sender: String,
    channel: String,
    state: Arc<Mutex<State>>,
}

impl TestEnv {
    pub fn new(data: impl ToString) -> Self {
        let mut config = Config::default();
        config.irc_config.name = "test_bot".into();
        config.irc_config.channels.push("#test_channel".into());
        config.irc_config.owners.push("test_owner".into());

        let mut state = State::default();
        state.insert(CachedConfig::new(config, "noye.toml"));

        Self {
            responder: YamlResponder::default(),
            data: data.to_string(),
            sender: "test_user".into(),
            channel: "#test_channel".into(),
            state: Arc::new(Mutex::new(state)),
        }
    }

    pub fn owner(mut self) -> Self {
        self.sender = "test_owner".into();
        self
    }

    pub fn user(mut self, user: impl ToString) -> Self {
        self.sender = user.to_string();
        self
    }

    pub fn channel(mut self, channel: impl ToString) -> Self {
        self.channel = channel.to_string();
        self
    }

    pub fn insert<T: 'static + Send + Sync>(self, item: T) -> Self {
        assert!(self.state.lock().now_or_never().unwrap().insert(item));
        self
    }

    pub fn config(self, f: impl FnOnce(&mut Config)) -> Self {
        {
            let mut state = self.state.lock().now_or_never().unwrap();
            let mut config = state
                .expect_get_mut::<CachedConfig>()
                .unwrap()
                .get_mut()
                .now_or_never()
                .unwrap()
                .unwrap();
            f(&mut config);
        }
        self
    }

    pub async fn execute<H, F>(self, handler: H) -> YamlResponder
    where
        H: Handler<YamlResponder, Fut = F>,
        F: std::future::Future<Output = anyhow::Result<()>> + Send + 'static,
        F::Output: Send + 'static,
    {
        let msg = crate::Message {
            sender: self.sender,
            channel: self.channel,
            data: self.data,
        };

        let (tx, rx) = tokio::sync::mpsc::channel(32);
        match handler
            .call(
                crate::Context {
                    args: Arc::new(msg),
                    writer: crate::Writer(tx),
                    state: self.state.clone(),
                    quit: Default::default(),
                },
                self.responder.clone(),
            )
            .await
        {
            Ok(..) => {}
            Err(err) if err.is::<crate::util::DontCareSigil>() => {}
            Err(err) => {
                let err = err
                    .chain()
                    .enumerate()
                    .fold(String::new(), |mut a, (i, err)| {
                        a.extend(format!("\n[{}] --> ", i).drain(..));
                        a.extend(err.to_string().drain(..));
                        a
                    });
                panic!("{}", err)
            }
        }

        self.responder
            .responses
            .lock()
            .unwrap()
            .extend(rx.map(Response::Raw).collect::<Vec<_>>().await);

        self.responder
    }
}
