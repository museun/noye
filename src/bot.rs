use crate::{AnyhowFut, Command, Config, Context, Handler, Irc, Prefix, Responder, State, Writer};

use futures::prelude::*;
use std::{collections::HashMap, future::Future, sync::Arc};
use tokio::sync::{Mutex, Notify};

pub struct CommandsMap<R> {
    map: HashMap<String, Arc<dyn Handler<R, Fut = AnyhowFut<'static>> + Send + 'static>>,
    _marker: std::marker::PhantomData<R>,
}

impl<R> Default for CommandsMap<R> {
    fn default() -> Self {
        Self {
            map: Default::default(),
            _marker: std::marker::PhantomData::default(),
        }
    }
}

impl<R: Responder + Send + 'static> CommandsMap<R> {
    pub fn add<H, F>(&mut self, cmd: impl ToString, handler: H) -> anyhow::Result<()>
    where
        H: Handler<R, Fut = F>,
        F: Future<Output = anyhow::Result<()>> + Send + 'static,
        F::Output: Send + 'static,
    {
        let cmd = cmd.to_string();
        if self.map.contains_key(&cmd) {
            anyhow::bail!("{} already exists as a command", cmd)
        }

        self.map
            .insert(cmd, Arc::new(move |state, resp| handler.call(state, resp)));

        Ok(())
    }
}

pub struct PassivesList<R> {
    list: Vec<Arc<dyn Handler<R, Fut = AnyhowFut<'static>> + Send + 'static>>,
    _marker: std::marker::PhantomData<R>,
}

impl<R> Default for PassivesList<R> {
    fn default() -> Self {
        Self {
            list: Default::default(),
            _marker: std::marker::PhantomData::default(),
        }
    }
}

impl<R: Responder + Send + 'static> PassivesList<R> {
    pub fn add<H, F>(&mut self, handler: H)
    where
        H: Handler<R, Fut = F>,
        F: Future<Output = anyhow::Result<()>> + Send + 'static,
        F::Output: Send + 'static,
    {
        self.list
            .push(Arc::new(move |state, resp| handler.call(state, resp)))
    }
}

pub struct Bot<R> {
    pub quit: Arc<Notify>,
    pub writer: Writer,
    pub commands: CommandsMap<R>,
    pub passives: PassivesList<R>,
    pub state: Arc<Mutex<State>>,
    _phantom: std::marker::PhantomData<R>,
}

impl<R: Responder + Send + 'static> Bot<R> {
    pub fn new(
        state: State,
        writer: Writer,
        commands: CommandsMap<R>,
        passives: PassivesList<R>,
    ) -> Self {
        let (quit, _phantom) = Default::default();
        Self {
            quit,
            commands,
            passives,
            writer,
            state: Arc::new(Mutex::new(state)),
            _phantom,
        }
    }

    pub async fn handle(&mut self, data: &str, responder: R) -> anyhow::Result<()> {
        let msg = Irc::parse(data)?;
        match msg.command {
            Command::Privmsg => {
                let context = Context::new(msg.into_message(), self);
                self.dispatch(context, responder.clone())
            }

            Command::Ping => {
                self.writer
                    .raw(format!("PONG {}", msg.data.unwrap()))
                    .await?;
            }

            Command::Ready => {
                let mut state = self.state.lock().await;
                let Config {
                    irc_config:
                        crate::config::Irc {
                            q_name,
                            q_pass,
                            channels,
                            ..
                        },
                    ..
                } = state.config().await?;

                match (q_name, q_pass) {
                    (Some(q_name), Some(q_pass)) if !q_name.is_empty() && !q_pass.is_empty() => {
                        self.writer
                            .raw(format!(
                                "PRIVMSG Q@CServe.quakenet.org :AUTH {} {}",
                                q_name, q_pass
                            ))
                            .await?;
                        self.writer.raw(format!("MODE {} +x", &msg.args[0])).await?;
                    }
                    _ => {
                        log::warn!("not authing with Q, joining channels");
                        for channel in channels {
                            self.writer.join(channel).await?;
                        }
                    }
                }
            }

            Command::Quit | Command::Nick => {
                let mut state = self.state.lock().await;
                let Config {
                    irc_config: crate::config::Irc { name, .. },
                    ..
                } = state.config().await?;

                let sender = match &msg.prefix {
                    Some(Prefix::User { nick, .. }) => nick,
                    _ => unreachable!(),
                };
                if name == sender {
                    self.writer.nick(name).await?;
                }
            }

            Command::NickCollision => {
                self.writer.nick(format!("{}_", msg.args[1])).await?;
            }

            Command::Invite => {
                self.writer.join(msg.data.unwrap()).await?;
            }

            Command::Numeric(396) => {
                let mut state = self.state.lock().await;
                let Config {
                    irc_config: crate::config::Irc { channels, .. },
                    ..
                } = state.config().await?;

                for channel in channels {
                    self.writer.join(channel).await?;
                }
            }

            _ => {}
        }
        Ok(())
    }

    fn dispatch(&self, context: Context, responder: R) {
        if context.args.data.starts_with('!') {
            let input = &context.args.data[1..];
            if let Some(head) = input.splitn(2, ' ').next() {
                if let Some(cmd) = self.commands.map.get(head) {
                    let head = head.to_string();
                    let fut =
                        cmd.call(context.clone(), responder.clone())
                            .inspect_err(move |err| {
                                crate::util::inspect_err(err, || format!("command '{}'", head))
                            });
                    tokio::task::spawn(fut);
                }
            }
        }

        for passive in &self.passives.list {
            let fut = passive
                .call(context.clone(), responder.clone())
                .inspect_err(move |err| crate::util::inspect_err(err, || "passive"));
            tokio::task::spawn(fut);
        }
    }
}
