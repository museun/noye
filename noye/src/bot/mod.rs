use crate::modules;
use prelude::*;
use std::sync::Arc;

/// Bot is the main item for creating a bot
pub struct Bot<T> {
    config: Config,
    client: crate::irc::Client<T>,
    dispatcher: crate::command::Dispatcher,
}

impl<T> Bot<T> {
    /// Create a bot from a provided [`Config`](../config/struct.Config.html) and a [`Client`](../irc/struct.Client.html)
    pub fn create(config: Config, client: crate::irc::Client<T>) -> Self {
        // this is when the bot started, so force the initialization here
        once_cell::sync::Lazy::force(&modules::builtin::START);

        let mut dispatcher = Dispatcher::default();
        modules::load_modules(&config, &mut dispatcher);
        Self {
            config,
            client,
            dispatcher,
        }
    }

    /// Get a mutable reference to the [`Dispatcher`](./prelude/struct.Dispatcher.html)
    ///
    /// This allows you to add:
    /// * [`Command handlers`](./prelude/struct.Dispatcher.html#method.command)
    ///     - for reacting to user-triggered commands
    ///     - ex: `!command`
    /// * [`Listening handlers`](./prelude/struct.Dispatcher.html#method.listener)
    ///     - for reacting to IRC events
    ///     - ex: `PING`
    /// * [`Passive handlers`](./prelude/struct.Dispatcher.html#method.passive)
    ///     - for reacting to regular expressions on user messages
    ///     - ex: `(^hello)\s(?P<what>.*?$)`
    pub fn dispatcher(&mut self) -> &mut crate::command::Dispatcher {
        &mut self.dispatcher
    }

    /// Run the bot until it disconnects
    pub async fn run_loop(mut self) -> anyhow::Result<()>
    where
        T: tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin,
    {
        let crate::config::Irc {
            nick, user, real, ..
        } = &self.config.irc_config;

        self.client
            .register(crate::irc::Registration {
                nick: &nick,
                user: &user,
                real: &real,
            })
            .await?;

        let config = Arc::new(self.config);

        loop {
            let msg = self.client.read().await?;
            log::trace!("{:?}", msg);
            for resp in self.dispatcher.dispatch(msg, Arc::clone(&config)).await {
                for line in resp.into_iter().filter(|s| !s.is_empty()) {
                    self.client.write(line).await?;
                }
            }
        }
    }
}

/// Module to allow for globbing all of the common imports used when writing modules/handlers
pub mod prelude;

mod template;
pub use template::Template;
