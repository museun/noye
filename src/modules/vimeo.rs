use crate::bot::prelude::*;
use crate::http;
use futures::prelude::*;

registry!("vimeo" => {
    passive => VIMEO_REGEX, hear_vimeo
});

static VIMEO_REGEX: &str = r"(?x)
    vimeo\.com/(?P<vid>\d+)
";

#[derive(Template, Debug)]
#[parent("vimeo")]
enum Output {
    Video {
        id: i64,
        width: i64,
        height: i64,
        duration: String,
        fps: String,
        title: String,
        owner: String,
    },
}

async fn hear_vimeo(context: Context, mut noye: Noye) -> anyhow::Result<()> {
    let vids = context.matches().get_many("vid")?;

    let client = http::new_client();
    let mut stream = concurrent_map("vimeo", None, vids, |vid| {
        let client = client.clone();
        async move { lookup(&client, vid).await }
    })
    .await;

    while let Some(resp) = stream.next().await {
        noye.say_template(&context, resp)?;
    }

    noye.nothing()
}

async fn lookup(client: &reqwest::Client, vid: &str) -> anyhow::Result<Output> {
    #[derive(Debug, serde::Deserialize)]
    struct Response {
        video: Video,
    }

    #[derive(Debug, serde::Deserialize)]
    struct Video {
        id: i64,
        width: i64,
        height: i64,
        duration: i64,
        fps: f64,
        title: String,
        owner: Owner,
    }

    #[derive(Debug, serde::Deserialize)]
    struct Owner {
        name: String,
    }

    let url = format!("https://player.vimeo.com/video/{}/config", vid);
    let video = crate::http::get_json(
        &client,
        &url,
        &crate::http::NoQuery,
        &[("Accept-Encoding", "identity")],
    )
    .map_ok(|resp: Response| resp.video)
    .await?;

    let output = Output::Video {
        id: video.id,
        width: video.width,
        height: video.height,
        duration: video.duration.as_timestamp(),
        fps: video.fps.to_string(),
        title: video.title,
        owner: video.owner.name,
    };
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bot::test::*;

    // TODO mock this

    #[test]
    #[ignore]
    fn hear_vimeo() {
        let tests = vec![
            (
                "https://vimeo.com/23960970",
                Output::Video {
                    id: 23_960_970,
                    width: 1280,
                    height: 720,
                    duration: "02:32".into(),
                    fps: "25".into(),
                    title: "NuFormer - 3D video mapping interactivity test. April 2011".into(),
                    owner: "NuFormer".into(),
                },
            ),
            (
                "https://vimeo.com/220883711",
                Output::Video {
                    id: 220_883_711,
                    width: 1920,
                    height: 1080,
                    duration: "03:11".into(),
                    fps: "25".into(),
                    title: "Mixed Reality - THEORIZ - RnD test 002".into(),
                    owner: "THÃ‰ORIZ".into(),
                },
            ),
            (
                "https://vimeo.com/16750764",
                Output::Video {
                    id: 16_750_764,
                    width: 1280,
                    height: 720,
                    duration: "02:40".into(),
                    fps: "24".into(),
                    title: "D7000 Video Test - Film Look".into(),
                    owner: "Melo".into(),
                },
            ),
        ];

        for (input, expected) in tests {
            let ctx = Context::mock_context_regex(input, VIMEO_REGEX);
            let resp = say_template(&ctx, expected);
            check(super::hear_vimeo, ctx, vec![&resp])
        }
    }
}
