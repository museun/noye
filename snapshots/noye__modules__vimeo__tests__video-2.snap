---
source: src/modules/vimeo.rs
expression: "resp.get_say::<responses::Vimeo>()"
---
Video:
  id: 220883711
  width: 1920
  height: 1080
  duration: "03:11"
  fps: "25"
  title: "[title]"
  owner: "[owner]"
